hamilton/lifecycle/default.py

```diff
@@ -584,10 +598,43 @@ def run_to_execute_node(
         # and truncate it/provide sentinels for every failure)
         # TODO -- decide what to do with collect
         """Executes a node. If the node fails, returns the sentinel value."""
-        for key, value in node_kwargs.items():
-            if value == self.sentinel_value:  # == versus is
-                return self.sentinel_value  # cascade it through
+        default_return = [self.sentinel_value] if is_expand else self.sentinel_value
+        _node_tags = future_kwargs["node_tags"]
+        can_inject = any(
+            _node_tags.get(k, "") == v for k, v in self.sentinel_injection_tags.items()
+        )
+        if not can_inject:
+            for key, value in node_kwargs.items():
+                if type(self.sentinel_value) is type(value):
+                    if self.sentinel_value == value:  # == versus is
+                        return default_return
+        if not is_expand:
```

Comment:
I don't know all the little details well enough to say for sure. The preceding parts are just supposed to handle if the node is skipped because an input node had an error.

The tests check what happens if the expand node gets an error sentinel, and goes to the collect and beyond, passing the sentinel. If it's not complex enough to detect what you're thinking of, I'm happy to update it.

Reply:
Yes, on second glance, I think you're right here, I was overly optimizing.